### PR TITLE
Sentry 1

### PR DESCRIPTION
**Pull request by bloodearnest**
Initial sentry support

 - wsgi handler
 - logging handler
 - flask support

### CODE REVIEW

File: docs/flask.rst [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+The sentry flask extension is configured to work with talisker.
+
+ * ```logging=False``` as Talisker has already set this up. This means the
+   other possible logging config is ignored.
+
+ * ```wrap_wsgi=False``` as Talisker has already set this up
+
+ * ```register_signal=True```, which is the default
+
+If for some reason you with to configure the flask sentry extension yourself::
```
Comment (ricardokirkner) [inline docs/flask.rst]: s/with/wish/ ?

File: docs/logging.rst [hunk 2]
```diff
@@ -195,12 +195,12 @@ line, following the `logfmt <https://brandur.org/logfmt>`_ idea. e.g.::
 
     * keys: any string, except:
         - `` ``, ``.``, and ``=`` are replaced by ``_``
-        - ``"`` is replaced by ````
+        - ``"`a` is replaced by ``""``
```
Comment (ricardokirkner) [inline docs/logging.rst]: is this correct? what is the significance of the letter a ?

File: docs/sentry.rst [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+======
+Sentry
+======
+
+Talisker provides out-of-the-box integration with sentry.
+
+Specifically, talisker adds:
+
+ * some default configuation of the sentry client
```
Comment (ricardokirkner) [inline docs/sentry.rst]: s/configuation/configuration/

File: docs/sentry.rst [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+ * sentry wsgi middleware
+ * sentry error log handler (for logged exception messages)
+ * log messge breadcrumbs (more breadcrumbs is a TODO)
```
Comment (ricardokirkner) [inline docs/sentry.rst]: s/messge/message/

File: docs/sentry.rst [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+To get the current sentry client, simply use::
+
+    talisker.sentry.get_client()
+
+Error Data
+----------
+
+Talisker configures sentry breadcrumbs for log messages at INFO or higher level.
+It also addes the request id as a tag to the sentry message.
```
Comment (ricardokirkner) [inline docs/sentry.rst]: s/addes/adds/

File: docs/sentry.rst [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+If you want to add some custom error context, you can use the client above as you
+would use the sentry client as normal.
+
+e.g.::
```
Comment (ricardokirkner) [inline docs/sentry.rst]: I'd collapse this into the previous paragraph, or alternatively use 'For example::' instead

File: docs/sentry.rst [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+your sentry client further via the usual config methods for those frameworks.
+
+If you wish to manually configure the sentry client, use the following::
+
+    talisker.sentry.set_client(**config)
+
+This will reconfigure and reset the sentry client used by the wsgi middleware
+and logging handler that Talisker sets up.
+
+Whichever way you wish to configure sentry, talisker will honour your
```
Comment (ricardokirkner) [inline docs/sentry.rst]: I think you always use Talisker capitalized, correct?

File: docs/statsd.rst [hunk 8]
```diff
@@ -13,7 +13,9 @@ Configuration
 Statsd can be configured by the STATSD_DSN envvar, patterned after the SENTRY_DSN.
 This combines all statsd config into a single DSN url.
 
-e.g.::
+e.g.
```
Comment (ricardokirkner) [inline docs/statsd.rst]: I'd collapse this into the previous paragraph, or alternatively use 'For example::' instead

File: talisker/flask.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
+# Copyright (C) 2016- Canonical Ltd
```
Comment (ricardokirkner) [inline talisker/flask.py]: you may want to update the year

File: talisker/flask.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+def _set_flask_sentry_client(app, **kwargs):
+    config = _get_flask_sentry_config(app)
+    config.update(kwargs)
+    # update the sentry client with the app config
+    logging.getLogger(__name__).info(
+        "updating sentry config with flask app configuration")
+    talisker.sentry.set_client(**config)
+
+
+def sentry(app, dsn=None, transport=None, **kwargs):
```
Comment (ricardokirkner) [inline talisker/flask.py]: given this wraps the Sentry class, why not call this function Sentry as well (as it's intended to be a simple wrapper)

File: talisker/sentry.py [hunk 11]
```diff
@@ -0,0 +1,146 @@
+# Copyright (C) 2016- Canonical Ltd
```
Comment (ricardokirkner) [inline talisker/sentry.py]: ditto about the year here

File: tests/test_flask.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
+# Copyright (C) 2016- Canonical Ltd
```
Comment (ricardokirkner) [inline tests/test_flask.py]: ditto

File: tests/test_sentry.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
+# Copyright (C) 2016- Canonical Ltd
```
Comment (ricardokirkner) [inline tests/test_sentry.py]: ditto

File: docs/logging.rst (hunk 2 above)
Comment (bloodearnest) [inline docs/logging.rst]: urgh, that's a typo

File: docs/flask.rst (hunk 1 above)
Comment (bloodearnest) [inline docs/flask.rst]: yep, will fix

File: docs/sentry.rst (hunk 6 above)
Comment (bloodearnest) [inline docs/sentry.rst]: done

File: docs/sentry.rst (hunk 7 above)
Comment (bloodearnest) [inline docs/sentry.rst]: yep

File: talisker/flask.py (hunk 9 above)
Comment (bloodearnest) [inline talisker/flask.py]: Yeah, I need to update all the copyright headers, will do that in a later PR

File: docs/statsd.rst (hunk 8 above)
Comment (bloodearnest) [inline docs/statsd.rst]: done

File: talisker/flask.py (hunk 10 above)
Comment (bloodearnest) [inline talisker/flask.py]: Well, it's not an object, so it looked wrong. But if you think that's best, I'd be ok, just seemed odd to me.

Comment (ricardokirkner) [inline talisker/flask.py]: ok, if it's not an object, but a plain function call, then I agree

**Pull request merged**